src/utils/work/fetchWork.js

```diff
@@ -1,56 +1,51 @@
 import { Environment, isBrowser } from 'utils/deploymentInfo';
 
-import { getBackendStatus } from 'redux/selectors';
-
 import cache from 'utils/cache';
-import generateETag from 'utils/work/generateETag';
-import { dispatchWorkRequest, seekFromS3 } from 'utils/work/seekWorkResponse';
-
-// Temporarily using gene expression without local cache
-const fetchGeneExpressionWorkWithoutLocalCache = async (
-  experimentId,
+import dispatchWorkRequest from 'utils/work/dispatchWorkRequest';
+import downloadFromS3 from 'utils/work/downloadFromS3';
+import waitForWorkRequest from './waitForWorkRequest';
+
+// retrieveData will try to get the data for the given experimentId and ETag from
+// the fastest source possible. It will try to get the data in order from:
+// 1. Browser cache
+// 2. S3 (cache) via signedURL
+// 3. Worker
+//   3.1. Via socket (small data)
+//   3.2. Via S3 (large data) via signedURL
+const retrieveData = async (experimentId,
+  ETag,
+  signedUrl,
+  request,
   timeout,
   body,
-  backendStatus,
-  environment,
-  broadcast,
-  extras,
-  dispatch,
-  getState,
-) => {
-  // If new genes are needed, construct payload, try S3 for results,
-  // and send out to worker if there's a miss.
-  const { pipeline: { startDate: qcPipelineStartDate } } = backendStatus;
+  dispatch) => {
+  // 1. Check if we have the ETag in the browser cache (no worker)
+  const cachedData = await cache.get(ETag);
+  if (cachedData) {
+    return cachedData;
+  }
 
-  const ETag = await generateETag(
-    experimentId,
-    body,
-    extras,
-    qcPipelineStartDate,
-    environment,
-    dispatch,
-    getState,
-  );
+  // 2. Check if data is cached in S3 so we can download from the signed URL (no worker)
+  if (signedUrl) {
```

Comment:
Is it possible to have the ETag in the cache, but not the signedURL? Can you give me an example of when that might happen? From the top of my head I am thinking that we'd always either have none of them or both, but I could be missing something

Reply:
yes, it can happen when:
*  the work results from S3 expire but the ones in the user browser have not.
* we have deleted manually or some other way the results from S3.

I am not sure if the expiration in the browser can be longer than S3. It shouldn't happen often.

Reply:
Makes sense, what you are saying sounds theoretically possible. `I am not sure if the expiration in the browser can be longer than S3. It shouldn't happen often.` <-- an example of this could be if the user has the experiment open on the browser for longer time than the S3 expiration time? Then the results in S3 will expire because they are never fetched, but the browser cache will not expire, given that the user checks the page in periods less than the browser cache expiration time

Reply:
in any case, the code is designed to handle that situation so I'm resolving this